In [1]:
import numpy as np
import torch

### Create the Data

In [2]:
# Scalar Tensor
x=torch.tensor(10,requires_grad=False)
print(x)

tensor(10)


### Mannual Cal

In [3]:
# cal another tensor using x
y= x**2
print(y)

tensor(100)


In [4]:
# cal derivative of y w.r.t x
# the darivative of x^2 -> 2x
darivative_of_y=2*x
print(darivative_of_y)

tensor(20)


### Auto Grad

In [5]:
# create the tensor 
a = torch.tensor(10.0,requires_grad=True)
print(a)
print(f"gradient stored in a = {a.grad}")

tensor(10., requires_grad=True)
gradient stored in a = None


In [6]:
# cal another tensor using a
b = a**2
print(f"b : {b}")

b : 100.0


In [7]:
# cal gradient Automatically
b.backward()

In [8]:
# check the gradients 
print(f"gradient stored in a = {a.grad}")

gradient stored in a = 20.0


In [9]:
# check the grad function
print(b.grad_fn)

In [10]:
c=a**3
c.backward()

In [11]:
print(a.grad)

tensor(320.)


### Note 

- Since a.grad alredy was having the value 20,
- The final answer is 320 -> 300+20=320 not 300
- This Called the Gradient accumlatuion

In [20]:
# if u want to clear the privious gradients
a.grad = torch.tensor(0.0)

In [23]:
d=a**3
m=d*2
m.backward()
print(d.grad_fn)